Update MPAS-O: 3d varying GM bolus and 2d varying phase speed

mark-petersen opened the pull request:
In the GM bolus calculation, there is a specified diffusivity and phase
speed (see Ferrari et al. 2010). Currently MPAS-O assumes these values
are fixed in time and space. This PR allows separate flags to enable
2D varying (+time) phase speed (based on the first baroclinic mode phase
speed) and a separate flag for 3D (+time) varying diffusivity.

See https://github.com/MPAS-Dev/MPAS-Model/pull/288

[NML]
[non-BFB]

mark-petersen commented:
@vanroekel this is a placeholder PR. The diffs won't be correct until MPAS-Dev/MPAS-Model#288 is merged in.

Based on the test results (see  MPAS-Dev/MPAS-Model#288), I think we should make vertically varying GM-Bolus default ON for all low-res runs. That would make this a non-BFB PR.

rljacob commented:
@mark-petersen,  I found a design document in the old Ocean Group space:
https://acme-climate.atlassian.net/wiki/spaces/OCNICE/pages/980713736/ocean+GM+Parameter+function+of+depth

But this v2 feature needs a code review document added here:
https://acme-climate.atlassian.net/wiki/spaces/ED/pages/36176475/Proposed+New+Features

And approval of the above by @golaz

vanroekel commented:
@rljacob the design doc is in place now.
https://acme-climate.atlassian.net/wiki/spaces/ED/pages/1023836492/W16+MPAS-Ocean+3D+varying+GM+bolus+kappa+Design+Document

vanroekel commented:
NOTE that this still needs the corresponding E3SM changes to add namelist flags and set defaults.  I'll try add my changes shortly

rljacob commented:
@vanroekel that design document needs to be within a larger code review document.  And "W16" was already taken so make it W17.

mark-petersen commented:
@vanroekel this is ready for flag updates. @stephenprice this is the branch to use for your test, after  Luke adds the namelist options.

mark-petersen commented:
FYI, juste rebased this on current master, so it should work on cori, but I haven't tested it.

vanroekel commented:
@stephenprice the branch is now ready for you to test.

mark-petersen commented:
Passes these tests on cori-knl:
```
PASS PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_intel COMPARE_base_modpes
PASS PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_intel COMPARE_base_single_thread
PASS PET_Ln9.T62_oEC60to30v3.GMPAS-IAF.cori-knl_intel COMPARE_base_single_thread
```
So I know the updated namelists work. Thanks @vanroekel and @jonbob

mark-petersen commented:
OK, I updated the design document here. Added performance test and reviewed the rest, and filled in the meta-data information at the top that is collected in the page one level up.
https://acme-climate.atlassian.net/wiki/spaces/ED/pages/1023836492/W17+MPAS-Ocean+3D+varying+GM+bolus+kappa+Design+Document

@jonbob let's proceed with this one.

mark-petersen commented:
@vanroekel please approve this PR.

@golaz, I think we are asking for team lead approval, and this falls under the water cycle. I think you just need to acknowledge that we filled out everything on the design document, see last post.

vanroekel commented on components/mpas-ocean/cime_config/buildnml:
```diff
@@ -819,6 +819,10 @@ def buildnml(case, caseroot, compname):
+	    lines.append('    <var name="kappaGM3D"/>')
+            lines.append('    <var name="cGMphaseSpeed"/>')
+	    lines.append('    <var name="velocityZonalTimesTemperature_GM"/>')
+            lines.append('    <var name="velocityMeridionalTimesTemperature_GM"/>')
```
could we also add GMBolusVelocityZonal and GMBolusVelocityMeridional here?

mark-petersen commented:
@jonbob I added those two variables, so all requests are completed.

jonbob commented:
@mark-petersen - what was that last commit? This PR doesn't say anything about bringing in a new grid and doesn't have the CIME support for it

mark-petersen commented:
Whoops - working on two things at once. I'll fix that, thanks for looking.

mark-petersen commented:
OK, just rolled back the last commit. This is the only thing I added for Luke's requested variables:
https://github.com/E3SM-Project/E3SM/pull/3057/commits/6ff481a1a92782bb4a93479fb474eae053e51894

jonbob commented:
merged to next

jonbob commented:
merged to master and expected DIFF's blessed
